### PR TITLE
Facility level reports

### PR DESCRIPTION
**Pull request by rsanheim**
**Story card:** https://app.clubhouse.io/simpledotorg/story/229/allow-new-district-reports-to-drill-into-facility-level-details

## Because

We want Facility level reports that match the new card based reports at the District (ie FacilityGroup) level. 

This basically reuses the same services for Facility level as we have for District. I am adding a new controller for now even though the code is mostly the same. I'd like to hold off on refactoring that a bit, as I think we should move the URL and controller naming to something like `/reports/[district_or_facility]/[slug]` ..but that will require some redirects to prevent old urls from breaking that early-access CVHOs are using.

## This addresses

Adding the new facility reports

## Screen Shots

image

image

### CODE REVIEW

File: app/models/facility.rb [hunk 1]
```diff
@@ -84,6 +84,11 @@ class Facility < ApplicationRecord
 
   friendly_id :name, use: :slugged
 
+  # For compatibility w/ parent FacilityGroups
+  def facilities
+    [self]
+  end
```
Comment (harimohanraj89) [inline app/models/facility.rb:90]: Love this.

File: app/services/region_report_service.rb [hunk 2]
```diff
@@ -19,7 +19,7 @@ def initialize(district:, selected_date:, current_user:)
 
   attr_reader :current_user
   attr_reader :data
-  attr_reader :district
+  attr_reader :region
```
Comment (harimohanraj89) [inline app/services/region_report_service.rb:22]: This change is really nice. Does it mean that, in the interrim, we can create some not-database-backed models like `Block` and `State` to be used by our reporting services? As long as they preserve the necessary interfaces for `name`, `facilities` `registered_patients`, etc?

Comment (rsanheim) [reply app/services/region_report_service.rb:22]: Yeah, we definitely could do that. I imagine that would search on our `block` or `state` text fields and return the related facilities?

Comment (harimohanraj89) [reply app/services/region_report_service.rb:22]: Following up on our server catchup conversation, yes that would be how we'd do that right away without any data model changes. Our state data is quite clean, so it may be a good next-one to try out. I'm not as confident about the block data.

**Comment by rsanheim** [comment]
> This looks great, very similar to what I had in mind to start accounting for Facilities/States/Blocks, etc.
> 
> Can we add some test coverage to the `RegionReportService`, ensuring it invokes the right sub-services?
> 
> Can we add some test coverage to the `ControlRateService`, ensuring it can calculate control rates for a single facility?

I added specs for both to cover the facility user case. I'm keeping these tests very functional level to make sure everything is working as we expect, and also so that they aren't too brittle. Once the interfaces settle down we could do more focused unit level testing at the right boundaries.
